packages/cardpay-sdk/sdk/reward-manager/base.ts

```diff
@@ -480,25 +477,65 @@ export default class RewardManager {
 The owner of reward safe ${safeAddress} is ${rewardSafeOwner}, but the signer is ${from}`
       );
     }
-    let weiAmount = new BN(amount);
-    if (weiAmount.gt(safeBalance)) {
-      throw new Error(
-        `Insufficient funds inside reward safe. safeBalance = ${fromWei(safeBalance)} and amount = ${fromWei(amount)}`
+
+    let estimate;
+    let withdrawPayload;
+    let weiAmount;
+    if (amount) {
+      //when amount is given, we assume that amount is string in wei that has already had gasCost deducted from it
+      let weiAmount = new BN(amount);
+      if (weiAmount.gt(safeBalance)) {
+        throw new Error(
+          `Insufficient funds inside reward safe. safeBalance = ${fromWei(safeBalance)} and amount = ${fromWei(amount)}`
+        );
+      }
+      let withdraw = await rewardSafeDelegate.methods.withdraw(rewardManagerAddress, tokenAddress, to, weiAmount);
+      withdrawPayload = withdraw.encodeABI();
+      estimate = await gasEstimate(
+        this.layer2Web3,
+        safeAddress,
+        rewardSafeDelegateAddress,
+        '0',
+        withdrawPayload,
+        Operation.DELEGATECALL,
+        tokenAddress
+      );
+      let gasCost = new BN(estimate.safeTxGas).add(new BN(estimate.baseGas)).mul(new BN(estimate.gasPrice));
```

Comment:
did we want to add a safety buffer for volatile gas tokens here?

Reply:
oops, i meant this comment for line 526

Reply:
Hmmmm sure I guess we can include the buffer. Tho there is this subjective question of what the buffer should be especially with different tokens. I was under the impression that this fix would already solve the issue without the buffer.

Reply:
the PR is more around improving DX so that callers don't have to think about gas estimation. but the underlying problem is still there.

Reply:
By putting all this logic in the sdk, we don't make the amount withdrawn dependent on a gasEstimate that might be lurking minutes before the transaction is executed. I think it DOES reduce the possiblity of the gas estimate deviation. 

We can add a buffer to reduce the possiblity even more.

Reply:
i mean the wallet was probably doing the gas estimation immediately before calling the SDK to do the withdraw, so it's more like you shave ms off the of delta between the estimation and the execution. but i think you shaved less than 1% off of the time that the price could have moved.

Reply:
That is to my point the question below.

- Before, we were doing two separate gas estimates. One in the wallet. One in the sdk. 
- Now, we are doing a single gas estimate. There is no delta in gas estimations at the moment. But we have to accept that the inital gas estimate is of a DIFFERENT withdraw amount than the "actual" (accounting for gas cost)

Reply:
ah i see what you are trying to say--having a buffer actually effects the parameters to the safe tx, which would mean that it has to be re-estimated, because the withdraw amount is different now (smaller after the buffer). which in turn adds more delay because of the extra round trip to the relay server. damned if you do and damned if you don't. We can forego the buffer for now and see how this plays out without that.

Reply:
let's go with this PR and we'll followup as necessary if we continue to see issues around this.

Reply:
I guess the buffer. For example adding gasCost= expectedGasCost+ buffer will reduce the likelihood of issues relayer erroring out.